### PR TITLE
Testbed for Stylus

### PR DESCRIPTION
**Pull request by dkebler**
…testing Stylus code with Dependents

With this you can test on your own.   I included a glob for testing as well.

### CODE REVIEW

File: tests/.deprc [hunk 1]
```diff
@@ -1,3 +1,4 @@
   {
-    "sass_root": "scss/" 
+    "sass_root": "scss/",
+    "stylus_root": "stylus/"
```
Comment (mrjoelkemp) [inline tests/.deprc]: `styles_root` now. It's meant to be generic enough to support all preprocessors. `styles` is super close to `stylus`, I know :)

**Comment by mrjoelkemp** [comment]
Any chance we could have a subset of the files that test the most important cases? Seems like quite a bit of files being added. Same goes for Sass; would love to keep the manual testing directories trim if we can.

**Comment by dkebler** [comment]
Made changes to .deprc works fine now both scss and stylus.  Was confused
about syntax for .deprc     Corrected pushed and made PR.

Also much better with allowing .deprc to be located in root of stylesheets
and yet function with window root at root of project!

Personally I like the test folders/files being more like a real project
(since that is where I extracted them from)   The base-theme tests three
deep.    So yea bunch of files but nothing in them really except imports.

When you say other test cases what more specifically did you mean/want?


On Mon, Jun 29, 2015 at 7:07 PM, Joel Kemp <notifications@github.com> wrote:

> Any chance we could have a subset of the files that test the most
> important cases? Seems like quite a bit of files being added. Same goes for
> Sass; would love to keep the manual testing directories trim if we can.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mrjoelkemp/Dependents/pull/147#issuecomment-116912812>
> .
>



-- 
Cheers, Cuidate, TTFN <http://www.youtube.com/watch?v=5Gu50vq5ux4>
David

**Comment by mrjoelkemp** [comment]
>  Corrected pushed and made PR.

Thanks!

> Also much better with allowing .deprc to be located in root of stylesheets and yet function with window root at root of project!

Yes. Should have implemented that a long time ago. Folks at work that have multiple repos open in a single window were always frustrated that they couldn't use the plugin.

> Personally I like the test folders/files being more like a real project (since that is where I extracted them from). 

I see the value in that. I'd prefer to have a single stylesheet that I go to where each import tests one of the following (might not be an exhaustive set):

1. a relatively pathed partial
1. importing a partial within a subdirectory
1. underscored partial on disk, but non-underscored partial
1. underscored on disk and in partial name 
1. with a file extension `"foo.scss"`
1. without a file extension `"foo"` 

This way, I just look at one file, click around and make sure all the key pieces still work. With a large number of files, I'll likely miss a few cases.

Will merge this soon and if you agree with minimizing the number of files, we could do that in a follow up.

**Comment by dkebler** [comment]
Yea leave the two test templates and I'll make a separate testbed of the
different flavors as in scss, sass, css, styl, and the various @import
differences in syntax with and without extensions, blah blah.    Won't get
to that until after the 4th weekend.

There is still the feature issue of "vendor" or bower components being
imported from outside the styles_root tree.  Not sure best way to address
this.  Maybe a styles_vendors_root in deprc which could be an array of
paths which could be checked if Dependents fails to find anything directly
below the styles_root.  That seems the most functional.    Could be same
for javascript as well as there are many javascript bower compenent.
To be really useful it might have walk down a tree looking for the file as
the full path might only be found in a gulp.js or bower.json file.

On Tue, Jun 30, 2015 at 11:33 AM, Joel Kemp <notifications@github.com>
wrote:

> Corrected pushed and made PR.
>
> Thanks!
>
> Also much better with allowing .deprc to be located in root of stylesheets
> and yet function with window root at root of project!
>
> Yes. Should have implemented that a long time ago. Folks at work that have
> multiple repos open in a single window were always frustrated that they
> couldn't use the plugin.
>
> Personally I like the test folders/files being more like a real project
> (since that is where I extracted them from).
>
> I see the value in that. I'd prefer to have a single stylesheet that I go
> to where each import tests one of the following (might not be an exhaustive
> set):
>
>    1. a relatively pathed partial
>    2. importing a partial within a subdirectory
>    3. underscored partial on disk, but non-underscored partial
>    4. underscored on disk and in partial name
>    5. with a file extension "foo.scss"
>    6. without a file extension "foo"
>
> This way, I just look at one file, click around and make sure all the key
> pieces still work. With a large number of files, I'll likely miss a few
> cases.
>
> Will merge this soon and if you agree with minimizing the number of files,
> we could do that in a follow up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mrjoelkemp/Dependents/pull/147#issuecomment-117294409>
> .
>



-- 
Cheers, Cuidate, TTFN <http://www.youtube.com/watch?v=5Gu50vq5ux4>
David

**Comment by mrjoelkemp** [comment]
> There is still the feature issue of "vendor" or bower components being imported from outside the styles_root tree.  Not sure best way to address this.

Yeah, 3rd party library support is definitely on the todo list. An array of js roots and a separate list of style roots is the route I'm favoring at the moment. 

I'm currently trying to make the underlying libs a bit more stable and dependency-updated (so that the npm dedupe actually detects shared dependency libraries across the various node tools powering the plugin). On top of stability, increasing the surface area of supported languages would help with adoption (at work, the PHP devs could totally use this plugin).

I'm thinking about seeking financial sponsorship for Dependents, which would allow me to focus my efforts on making the tool more useful on nights and weekends. Not sure of the best way of doing this; any ideas are appreciated.

**Comment by dkebler** [comment]
I've got a start on learning programming in javascript and would like to
use Dependents as a "real world" project to cut my teeth on.   So I would
like to work on this adding 3rd party (vendor) support.

To this end (learning javascript and python) I can also work on adding new
language support (e.g. php, ruby, python, go, various html templating
languages).   If your underlying code is written appropriately shouldn't
adding new language support be a pretty straight forward affair?   Maybe
work (code/documentation) should be done to make it possible for folks to
make a "plugin" for dependents so they can add their own language without
much trouble.

Could we set up some place where I can ask some questions about the
organization of your code and maybe a few pointers and advice (like the
best learning resources).  If we keep this conversation "organized" then
should be useful to others wanting to contribute down the line.   How about
a private ghost blog?  Ghost is a node.js blog solution so seems
appropriate.  I could run a copy at an AWS VM.

As to financial support have you considered a "kickstarter" or the like?
You have a working product which is a plus.  Having some financial backing
could mean expanded support for Atom editor (which I guess is pure
javascript (no python).  Albiet it's only like 3% of users as of 2014.
https://blog.codeanywhere.com/most-popular-ides-code-editors/.    Other
editors?  Eclipse???   Of course expanded (plugin?) support for more
languages.



On Wed, Jul 1, 2015 at 6:57 AM, Joel Kemp <notifications@github.com> wrote:

> Merged #147 <https://github.com/mrjoelkemp/Dependents/pull/147>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mrjoelkemp/Dependents/pull/147#event-345439133>.
>



-- 
Cheers, Cuidate, TTFN <http://www.youtube.com/watch?v=5Gu50vq5ux4>
David

**Comment by mrjoelkemp** [comment]
> If your underlying code is written appropriately shouldn't adding new language support be a pretty straight forward affair?

The code gets more appropriate as I see the pain points of adding new languages. It's my goal that adding a new language should be as simple as implementing a new detective and the module lookup algorithm and plugging those into factories like https://github.com/mrjoelkemp/node-precinct so that the modules that power features (like https://github.com/mrjoelkemp/node-dependents) don't need to worry at all about the file type or JS module type; it'll be file agnostic and still work.

> Maybe work (code/documentation) should be done to make it possible for folks to make a "plugin" for dependents so they can add their own language without much trouble.

As stated above, the workflow isn't quite there yet. When I get around to adding Less support, I'll create the abstractions and do further refactoring to get where I need it to be. 
 
> So I would like to work on this adding 3rd party (vendor) support.

Is that the #155 issue?

> Could we set up some place where I can ask some questions about the organization of your code and maybe a few pointers and advice (like the best learning resources). 

Yes. Absolutely. https://gitter.im/mrjoelkemp/Dependents. There's an issue #148 to add that to the Readme so others can ask questions.

> As to financial support have you considered a "kickstarter" or the like?

Great idea. I'll consider it and possibly get one going in the next couple of weeks.

**Pull request merged**